src/iter.rs

```diff
@@ -53,47 +56,51 @@ where
+        let mut path = P::default();
 
         loop {
-            return match M::path(self.state.iter().copied(), &mut path, self.separator) {
-                // Not having consumed any name/index, the only possible case here is a bare option.
-                // And that can not return `NotFound`.
-                Err(Error::NotFound(0)) => unreachable!(),
-                // Iteration done
+            return match M::path(self.state, &mut path, self.separator) {
+                // Out of valid indices at the root: iteration done
                 Err(Error::NotFound(1)) => {
                     debug_assert_eq!(self.count.unwrap_or_default(), 0);
                     None
                 }
-                // Node not found at depth: reset current index, increment parent index, then retry
-                Err(Error::NotFound(depth)) => {
-                    path = Self::Item::default();
+                // Node not found at depth: reset current index, increment parent index,
+                // then retry path()
+                Err(Error::NotFound(depth @ 2..)) => {
+                    path = P::default();
                     self.state[depth - 1] = 0;
                     self.state[depth - 2] += 1;
                     continue;
                 }
-                // Iteration done for a bare Option
-                Ok(Ok::Leaf(0)) if self.count == Some(0) => None,
-                // Root is `Leaf`: bare Option.
+                // Found a leaf at the root: bare Option
                 // Since there is no way to end iteration by triggering `NotFound` on a bare Option,
-                // we force the count to Some(0) and trigger on that (see above).
-                Ok(Ok::Leaf(0)) => {
-                    debug_assert_eq!(self.count.unwrap_or(1), 1);
-                    self.count = Some(0);
-                    Some(path)
+                // we force the count to Some(0) and trigger on that.
+                Ok(0) => {
+                    if self.count == Some(0) {
+                        None
+                    } else {
+                        debug_assert_eq!(self.count.unwrap_or(1), 1);
+                        self.count = Some(0);
+                        Some(Ok(path))
+                    }
                 }
-                // Non-root leaf: advance at current depth
-                Ok(Ok::Leaf(depth)) => {
+                // Non-root leaf: advance index at current depth
+                Ok(depth) => {
                     self.count = self.count.map(|c| c - 1);
                     self.state[depth - 1] += 1;
-                    Some(path)
+                    Some(Ok(path))
                 }
                 // If we end at a leaf node, the state array is too small.
-                Ok(Ok::Internal(_depth)) => panic!("Path iteration state too small"),
-                Err(e) => panic!("Path iteration: {e:?}"),
+                Err(e @ (Error::TooShort(_) | Error::Inner(_))) => Some(Err(e)),
```

Comment:
Isn't this guaranteed to not happen because we check `L` against the max depth specified by `metadata`? Seems like we could internally panic instead of forcing out a `Result` that the user has to unwrap that should never be `Err()` if miniconf is working properly.

Reply:
It's a choice. If you do `iter_paths()` (`PathIter::new()`) then it checks and errors early.
If you do `iter_paths_unchecked()` (`PathIter::new_unchecked()`), you'll get `TooShort()` when iterating (before it just paniced).
In the mqtt client because of other constraints we check metadata explicitly, then use `_unchecked()` and unwrap in the iteration.

Reply:
The intent of `PathIter::new_unchecked()` was that the operator of `miniconf` **guaranteed** that the state vector was long enough, which is why miniconf panicked internally for ergonomics. I'm not personally a fan of this API change when I don't see any utility in it

Reply:
I don't see it that wal.

In the worst case this is moving panics (string too short, state too short) in a library to an unwrap (thus panic) in the application (or a higher layer as mqtt-client). There is no penalty for that.
In the best case these errors can now be handled in the application.

`_unchecked()` avoids gathering the complete metadata for each iteration an does not do the early check for state length.

Reply:
The original design intent was that miniconf iteration would never be error-able. This change is reversing that design choice, and I don't really see why

Reply:
I guess my main point is that this introduces thrash to the end user and the miniconf code base, and I don't really see the purpose

Reply:
Before the iteration could always panic but would never return an error.
There was support for checks ahead of time (still at runtime) to exclude these panics. Then the errors would arise before iterating. Now iteration never panics.
The checks are still there (the early string length check is in mqtt-client). The only change to the before-iteration checks is that the capacity check needs to be done by the user (since `Write` doesn't give capacity info from the type alone, unlike `heapless::String<N>`). On balance that change (moving the capacity check) should be clearly a positive one.
When choosing between an API that panics (before) and one that errors (now), I'd also prefer the later as that at least in principle allows handling the error.

Reply:
I agree that when converting to the new API, adding an `.unwrap()` to at the callsite adds some thrashing. But it is a gain in robustness. And AFAICT it's no meaningful difference in generated code in the current practical cases. For the user right now that only uses `MqttClient` there is zero change anyway.